File: codeX-frontend/src/services/native/filesystem.js
/**
 * Native Filesystem Service
 * Bridges React UI to local filesystem via Tauri IPC on desktop,
 * while maintaining fallback to File System Access API on web.
 */

import { isDesktopApp } from './platform';
import { localFileSystem } from '../localFileSystem';

export const filesystemService = {
  isSupported() {
    return isDesktopApp() || localFileSystem.isSupported();
  },

  async pickFolder() {
    if (isDesktopApp()) {
      const { invoke } = await import('@tauri-apps/api/core');
      const folderPath = await invoke('fs_pick_folder');
      if (!folderPath) return null;

      const entries = await invoke('fs_list_directory', { path: folderPath, maxDepth: 6 });
      return {
        path: folderPath,
        name: folderPath.replace(/\\/g, '/').split('/').pop() || 'Project',
        entries,
      };
    }

    // Web Fallback: File System Access API
    const result = await localFileSystem.openDirectory();
    return {
      path: null,
      name: result.folderName,
      entries: result.tree,
      handle: result.dirHandle,
      fileHandles: result.fileHandles,
      dirHandles: result.dirHandles,
    };
  },

  async pickFile() {
    if (isDesktopApp()) {
      const { invoke } = await import('@tauri-apps/api/core');
      return await invoke('fs_pick_file');
    }
    return null;
  },

  async saveFileDialog(defaultName) {
    if (isDesktopApp()) {
      const { invoke } = await import('@tauri-apps/api/core');
      return await invoke('fs_save_file_dialog', { defaultName });
    }
    return null;
  },

  async listDirectory(path, maxDepth = 6) {
    if (isDesktopApp()) {
      const { invoke } = await import('@tauri-apps/api/core');
      return await invoke('fs_list_directory', { path, maxDepth });
    }
    return [];
  },

  async readFile(path) {
    if (isDesktopApp()) {
      const { invoke } = await import('@tauri-apps/api/core');
      return await invoke('fs_read_file', { path });
    }
    throw new Error('Direct path reading is only available on desktop');
  },

  async writeFile(path, content) {
    if (isDesktopApp()) {
      const { invoke } = await import('@tauri-apps/api/core');
      return await invoke('fs_write_file', { path, content });
    }
    throw new Error('Direct path writing is only available on desktop');
  },

  async createFile(path) {
    if (isDesktopApp()) {
      const { invoke } = await import('@tauri-apps/api/core');
      return await invoke('fs_create_file', { path });
    }
    throw new Error('Direct file creation is only available on desktop');
  },

  async createFolder(path) {
    if (isDesktopApp()) {
      const { invoke } = await import('@tauri-apps/api/core');
      return await invoke('fs_create_folder', { path });
    }
    throw new Error('Direct folder creation is only available on desktop');
  },

  async rename(oldPath, newPath) {
    if (isDesktopApp()) {
      const { invoke } = await import('@tauri-apps/api/core');
      return await invoke('fs_rename', { oldPath, newPath });
    }
    throw new Error('Direct rename is only available on desktop');
  },

  async delete(path) {
    if (isDesktopApp()) {
      const { invoke } = await import('@tauri-apps/api/core');
      return await invoke('fs_delete', { path });
    }
    throw new Error('Direct deletion is only available on desktop');
  },

  async exists(path) {
    if (isDesktopApp()) {
      const { invoke } = await import('@tauri-apps/api/core');
      return await invoke('fs_exists', { path });
    }
    return false;
  },

  async watchDirectory(path, onChange) {
    if (!isDesktopApp()) return () => {};

    const { invoke } = await import('@tauri-apps/api/core');
    const { listen } = await import('@tauri-apps/api/event');

    await invoke('fs_watch', { path });
    const unlisten = await listen('fs-change', (event) => {
      if (onChange) {
        onChange(event.payload);
      }
    });

    return async () => {
      unlisten();
      try {
        await invoke('fs_unwatch');
      } catch {}
    };
  },
};

File: codeX-frontend/src-tauri/src/commands/filesystem.rs
use crate::filesystem::service::{self, FileEntry};
use crate::filesystem::watcher::FileWatcher;
use crate::platform;
use std::path::Path;
use tauri::{AppHandle, State};

#[tauri::command]
pub fn fs_get_default_workspace_dir() -> String {
    service::get_default_workspace_dir()
}

#[tauri::command]
pub fn fs_create_project_folder(project_name: String, parent_path: Option<String>) -> Result<String, String> {
    service::create_project_folder(&project_name, parent_path)
}

#[tauri::command]
pub fn fs_pick_folder() -> Option<String> {
    service::pick_folder()
}

#[tauri::command]
pub fn fs_pick_file() -> Option<String> {
    service::pick_file()
}

#[tauri::command]
pub fn fs_save_file_dialog(default_name: Option<String>) -> Option<String> {
    service::save_file_dialog(default_name)
}

#[tauri::command]
pub fn fs_list_directory(path: String, max_depth: Option<usize>) -> Result<Vec<FileEntry>, String> {
    service::list_directory_flat(&path, max_depth.unwrap_or(6))
}

#[tauri::command]
pub fn fs_read_file(path: String) -> Result<String, String> {
    service::read_file(&path)
}

#[tauri::command]
pub fn fs_write_file(path: String, content: String) -> Result<(), String> {
    service::write_file(&path, &content)
}

#[tauri::command]
pub fn fs_create_file(path: String) -> Result<(), String> {
    service::create_file(&path)
}

#[tauri::command]
pub fn fs_create_folder(path: String) -> Result<(), String> {
    service::create_folder(&path)
}

#[tauri::command]
pub fn fs_rename(old_path: String, new_path: String) -> Result<(), String> {
    service::rename(&old_path, &new_path)
}

#[tauri::command]
pub fn fs_delete(path: String) -> Result<(), String> {
    service::delete(&path)
}

#[tauri::command]
pub fn fs_exists(path: String) -> bool {
    service::exists(&path)
}

#[tauri::command]
pub fn fs_watch(
    app: AppHandle,
    watcher_state: State<'_, FileWatcher>,
    path: String,
) -> Result<(), String> {
    watcher_state.watch(app, &path)
}

#[tauri::command]
pub fn fs_unwatch(watcher_state: State<'_, FileWatcher>) -> Result<(), String> {
    watcher_state.unwatch();
    Ok(())
}

#[tauri::command]
pub fn fs_reveal_in_file_manager(path: String) -> Result<(), String> {
    platform::reveal_in_file_manager(Path::new(&path))
}
